File: daemon/src/features/ledger/say.ts
import type { CommitTouch } from '@daemon/types/git'
import type { Actor, LedgerEntry } from '@daemon/types/ledger'

/**
 * The ledger in words, for whoever is reading it — a tool answering an agent, and the line
 * under a document in the doc pane.
 *
 * Past tense throughout, and "as part of" for an errand rather than "wrote": a row says what
 * was true when it was written and nothing has told the ledger since, and an errand's row
 * says which errand a file was part of, never which line was whose.
 */
export function sayAct(entry: LedgerEntry, now = Date.now()): string {
  const when = ago(now - entry.at)
  const said = [`${who(entry.actor)} ${did(entry)} ${when}`]
  if (entry.actor.context) said.push(`in ${entry.actor.context}`)
  return said.join(', ')
}

/** What git has to say, which is a different question and is labelled as one wherever it is
 *  shown: a commit is when work was filed and by whichever author was configured. */
export function sayCommit(touch: CommitTouch, now = Date.now()): string {
  const at = Date.parse(touch.at)
  const when = Number.isNaN(at) ? touch.at : ago(now - at)
  const sha = touch.sha.slice(0, 7)
  return `git: last committed by ${touch.author} ${when} — “${touch.subject}” (${sha})`
}

function who(actor: Actor): string {
  switch (actor.kind) {
    case 'agent': {
      const badge = [actor.persona, actor.model].filter(Boolean).join(', ')
      return `${actor.name ?? 'an agent'} (agent${badge ? `, ${badge}` : ''})`
    }
    case 'chat':
      return 'the page’s chat'
    case 'task':
      return `a task run${actor.id ? ` (${actor.id})` : ''}`
    case 'person':
      return 'somebody typing in the editor'
    case 'unknown':
      return 'somebody the app could not name'
  }
}

function did(entry: LedgerEntry): string {
  switch (entry.action) {
    case 'write':
      return entry.created ? 'made this' : 'changed this'
    case 'move':
      return entry.note ? `moved this here from ${entry.note}` : 'moved this here'
    case 'delete':
      return 'deleted this'
    case 'errand':
      return entry.note
        ? `changed this as part of “${entry.note}”`
        : 'changed this in an errand'
    case 'commit':
      return 'committed this'
  }
}

/** How long ago, in the roundest unit that still says something. */
export function ago(ms: number): string {
  if (ms < 60_000) return 'just now'
  const minutes = Math.round(ms / 60_000)
  if (minutes < 60) return `${String(minutes)} minute${minutes === 1 ? '' : 's'} ago`
  const hours = Math.round(minutes / 60)
  if (hours < 24) return `${String(hours)} hour${hours === 1 ? '' : 's'} ago`
  const days = Math.round(hours / 24)
  return `${String(days)} day${days === 1 ? '' : 's'} ago`
}

File: daemon/internal/ledger/say.go
// The ledger in words, for whoever is reading it — a tool answering an agent, and the line under a
// document in the doc pane.
//
// Past tense throughout, and "as part of" for an errand rather than "wrote": a row says what was
// true when it was written and nothing has told the ledger since, and an errand's row says which
// errand a file was part of, never which line was whose.

package ledger

import (
	"strconv"
	"strings"
	"time"

	"github.com/broodmotherai/broodmother/daemon/internal/git"
)

func SayAct(entry Entry, now int64) string {
	said := who(entry.Actor) + " " + did(entry) + " " + Ago(now-entry.At)
	if entry.Actor.Context != "" {
		said += ", in " + entry.Actor.Context
	}
	return said
}

// SayCommit is what git has to say, which is a different question and is labelled as one wherever
// it is shown: a commit is when work was filed and by whichever author was configured.
func SayCommit(touch git.CommitTouch, now int64) string {
	when := touch.At
	if at, ok := millisOf(touch.At); ok {
		when = Ago(now - at)
	}
	sha := touch.Sha
	if len(sha) > 7 {
		sha = sha[:7]
	}
	return "git: last committed by " + touch.Author + " " + when + " — “" + touch.Subject + "” (" + sha + ")"
}

func who(actor Actor) string {
	switch actor.Kind {
	case AgentActor:
		name := actor.Name
		if name == "" {
			name = "an agent"
		}
		badge := []string{}
		if actor.Persona != "" {
			badge = append(badge, actor.Persona)
		}
		if actor.Model != "" {
			badge = append(badge, actor.Model)
		}
		if len(badge) == 0 {
			return name + " (agent)"
		}
		return name + " (agent, " + strings.Join(badge, ", ") + ")"
	case ChatActor:
		return "the page’s chat"
	case TaskActor:
		if actor.ID != "" {
			return "a task run (" + actor.ID + ")"
		}
		return "a task run"
	case PersonActor:
		return "somebody typing in the editor"
	default:
		return "somebody the app could not name"
	}
}

func did(entry Entry) string {
	switch entry.Action {
	case Write:
		if entry.Created != nil && *entry.Created {
			return "made this"
		}
		return "changed this"
	case Move:
		if entry.Note != "" {
			return "moved this here from " + entry.Note
		}
		return "moved this here"
	case Delete:
		return "deleted this"
	case Errand:
		if entry.Note != "" {
			return "changed this as part of “" + entry.Note + "”"
		}
		return "changed this in an errand"
	case Commit:
		return "committed this"
	}
	return string(entry.Action)
}

// Ago is how long ago, in the roundest unit that still says something.
func Ago(ms int64) string {
	if ms < 60_000 {
		return "just now"
	}
	minutes := rounded(float64(ms) / 60_000)
	if minutes < 60 {
		return strconv.Itoa(minutes) + " minute" + plural(minutes) + " ago"
	}
	hours := rounded(float64(minutes) / 60)
	if hours < 24 {
		return strconv.Itoa(hours) + " hour" + plural(hours) + " ago"
	}
	days := rounded(float64(hours) / 24)
	return strconv.Itoa(days) + " day" + plural(days) + " ago"
}

func plural(count int) string {
	if count == 1 {
		return ""
	}
	return "s"
}

// millisOf reads git's ISO 8601. Not a time it can read is shown as it stands rather than as a
// duration from a date nobody could parse.
func millisOf(at string) (int64, bool) {
	for _, layout := range []string{time.RFC3339, "2006-01-02T15:04:05Z0700", time.RFC3339Nano} {
		if held, err := time.Parse(layout, at); err == nil {
			return held.UnixMilli(), true
		}
	}
	return 0, false
}

// rounded is JavaScript's Math.round: a half goes up, including a negative half.
func rounded(value float64) int {
	if value < 0 {
		return -int(-value + 0.5)
	}
	return int(value + 0.5)
}
